vendor.conf

```diff
@@ -11,7 +11,8 @@ github.com/coreos/go-semver             8ab6407b697782a06568d4b7f1db25550ec2e4c6
 github.com/deckarep/golang-set          ef32fa3046d9f249d399f98ebaf9be944430fd1d
 go.etcd.io/bbolt                        7ee3ded59d4835e10f3e7d0f7603c42aa5e83820 # v1.3.1-etcd.8
 
-github.com/docker/docker                dbe4a30928d418e0570891a09703bcbc0e4997a1
+github.com/docker/docker                5f0231bca193320e1a3d785a3ade0e64241fe580
+github.com/containerd/containerd        bd46ea519173cc28445da9b2a031b6139e18eb05
```

Comment:
Do you know which part of the code caused the problem? Wondering if we can refactor the code to prevent that we have to add containerd as a new dependency.

Reply:
The following code can cause compilation problems on mips.

diff --git a/vendor/github.com/docker/docker/pkg/system/stat_linux.go b/vendor/github.com/docker/docker/pkg/system/stat_linux.go
index 98c9eb1..17d5d13 100644
--- a/vendor/github.com/docker/docker/pkg/system/stat_linux.go
+++ b/vendor/github.com/docker/docker/pkg/system/stat_linux.go
@@ -8,7 +8,8 @@ func fromStatT(s *syscall.Stat_t) (*StatT, error) {
                mode: s.Mode,
                uid:  s.Uid,
                gid:  s.Gid,
-               rdev: s.Rdev,
+               // the type is 32bit on mips
+               rdev: uint64(s.Rdev), // nolint: unconvert
                mtim: s.Mtim}, nil
 }

Reply:
The following code can cause compilation problems on mips.
`diff --git a/vendor/github.com/docker/docker/pkg/system/stat_linux.go b/vendor/github.com/docker/docker/pkg/system/stat_linux.go
index 98c9eb1..17d5d13 100644
--- a/vendor/github.com/docker/docker/pkg/system/stat_linux.go
+++ b/vendor/github.com/docker/docker/pkg/system/stat_linux.go
@@ -8,7 +8,8 @@ func fromStatT(s *syscall.Stat_t) (*StatT, error) {
                mode: s.Mode,
                uid:  s.Uid,
                gid:  s.Gid,
-               rdev: s.Rdev,
+               // the type is 32bit on mips
+               rdev: uint64(s.Rdev), // nolint: unconvert
                mtim: s.Mtim}, nil
}`

Reply:
> diff --git a/vendor/github.com/docker/docker/pkg/system/stat_linux.go b/vendor/github.com/docker/docker/pkg/system/stat_linux.go
index 98c9eb1..17d5d13 100644
--- a/vendor/github.com/docker/docker/pkg/system/stat_linux.go
+++ b/vendor/github.com/docker/docker/pkg/system/stat_linux.go
@@ -8,7 +8,8 @@ func fromStatT(s *syscall.Stat_t) (*StatT, error) {
                mode: s.Mode,
                uid:  s.Uid,
                gid:  s.Gid,
-               rdev: s.Rdev,
+               // the type is 32bit on mips
+               rdev: uint64(s.Rdev), // nolint: unconvert
                mtim: s.Mtim}, nil
}

Reply:
```
diff --git a/vendor/github.com/docker/docker/pkg/system/stat_linux.go b/vendor/github.com/docker/docker/pkg/system/stat_linux.go
index 98c9eb1..17d5d13 100644
--- a/vendor/github.com/docker/docker/pkg/system/stat_linux.go
+++ b/vendor/github.com/docker/docker/pkg/system/stat_linux.go
@@ -8,7 +8,8 @@ func fromStatT(s *syscall.Stat_t) (*StatT, error) {
                mode: s.Mode,
                uid:  s.Uid,
                gid:  s.Gid,
-               rdev: s.Rdev,
+               // the type is 32bit on mips
+               rdev: uint64(s.Rdev), // nolint: unconvert
                mtim: s.Mtim}, nil
}
```

Reply:
When I imported the new package github.com/docker/docker/errdefs,
the file vendor/github.com/docker/docker/errdefs/http_helpers.go
refered to github.com/containerd/containerd/errdefs
```
--- a/vendor/github.com/docker/docker/errdefs/http_helpers.go
+++ b/vendor/github.com/docker/docker/errdefs/http_helpers.go
@@ -4,6 +4,7 @@ import (
        "fmt"
        "net/http"

+       containerderrors "github.com/containerd/containerd/errdefs"
        "github.com/docker/distribution/registry/api/errcode"
        "github.com/sirupsen/logrus"
        "google.golang.org/grpc/codes"
```

Reply:
Thanks! Ok, so the first one, I'd have to have a look into.

It looks like the second one (`docker/docker/errdefs`) looks to be only used in tests, and looks like we could easily  replace with a local helper. That would at least prevent the containerd package from being added as dependency/

Let me give it a quick try and open a PR for that.

Reply:
This is what I came with for the first one;

`fromStatT()` is used by:

- ~docker/docker/pkg/system.FromStatT()~ (is unused)
- ~docker/docker/pkg/system.Lstat()~ (is unused)
- docker/docker/pkg/system.Stat()

`docker/docker/pkg/system.Stat()` is used by:

- ~docker/docker/pkg/idtools.CanAccess()~ (is unused)
- ~docker/docker/pkg/idtools.CanAccess()~ (is unused)
- docker/docker/pkg/idtools.mkdirAs()


`docker/docker/pkg/idtools.mkdirAs()` is used by:

- ~docker/docker/pkg/idtools.MkdirAllAndChown()~ (is unused)
- ~docker/docker/pkg/idtools.MkdirAllAndChownNew()~ (is unused)
- ~docker/docker/pkg/idtools.MkdirAndChown()~ (is unused)

So, it's not actually used in this code _but_ the `github.com/docker/docker/pkg/idtools` package is _imported_ by `github.com/docker/docker/pkg/homedir`, which is used in a single place;

libnetwork/client/mflag/flag.go imports github.com/docker/docker/pkg/homedir, and uses:

- `homedir.Get()`
- `homedir.GetShortcutString()`

https://github.com/moby/libnetwork/blob/b10559f6c05e73cc34b7221784ef09f6cf2c9a7a/client/mflag/flag.go#L541

https://github.com/moby/libnetwork/blob/b10559f6c05e73cc34b7221784ef09f6cf2c9a7a/client/mflag/flag.go#L564

This one may be _slightly_ more work to replace (need to look into how critical the full behavior is), but should probably be possible.

Reply:
opened https://github.com/moby/libnetwork/pull/2561 for the first one

Reply:
and initial work for the second one in https://github.com/moby/libnetwork/pull/2562

Reply:
Sorry，I think that the github.com/docker/docker/pkg/idtools package is not imported by github.com/docker/docker/pkg/homedir.
Please check it again.

Reply:
And I cannot understand the ultimate purpose of your modification. Do you want to delete the function fromStatT completely?

Reply:
> Sorry，I think that the github.com/docker/docker/pkg/idtools package is not imported by github.com/docker/docker/pkg/homedir. Please check it again.

I saw it was imported here; https://github.com/moby/libnetwork/blob/7f60f05552dd6f88aa0a36640186fe155c471b82/vendor/github.com/docker/docker/pkg/homedir/homedir_linux.go#L9-L19

> And I cannot understand the ultimate purpose of your modification. Do you want to delete the function fromStatT completely?

No, my main goal was to try to reduce the dependency graph. As a "side-effect", if those dependencies are no longer needed, the  `fromStatT` function would no longer be used. 

But looks like we may not be able to do so directly (the linked PR's are a _start_)